File: account/src/main/java/br/com/accenture/wallet/account/service/AccountService.java
package br.com.accenture.wallet.account.service;

import br.com.accenture.wallet.account.domain.*;
import br.com.accenture.wallet.account.repository.AccountRepository;
import net.jodah.failsafe.Failsafe;
import net.jodah.failsafe.RetryPolicy;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.context.annotation.ApplicationScope;
import java.net.ConnectException;
import java.time.Duration;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;
import static java.util.Objects.isNull;

@Service
@ApplicationScope
public class AccountService {
    private final Logger logger = LoggerFactory.getLogger(AccountService.class);

    private final AccountRepository accountRepository;

    private final CustomerClient customerClient;

    @Autowired
    public AccountService(AccountRepository repo, CustomerClient client) {
        this.accountRepository = repo;
        this.customerClient = client;
    }

    public List<AccountModel> findAll() {
        return accountRepository.findAll().stream()
            .map(AccountModel::fromEntity)
            .collect(Collectors.toList());
    }

    @Async
    public CompletableFuture<AccountModel> create(AccountModel model) {
        return findCustomer(model.getCustomer()).thenApply(customer -> {
            if (isNull(customer)) {
                logger.warn("Customer nao encontrado com ID {}", model.getCustomer());
                return null;
            }

            logger.info("Criando uma conta para o cliente ".concat(customer.getName()));
            final Account account = accountRepository.save(model.toEntity()
                .setBalance(0.0)
                .setTransactionLimit(100.0));
            model.setId(account.getId());
            return model;
        });
    }

    public Optional<AccountModel> getById(String value) {
        final Optional<Account> account = accountRepository.findById(value);
        if (account.isEmpty()) return Optional.empty();
        return Optional.of(AccountModel.fromEntity(account.get()));
    }

    public void remove(String accountId) {
        accountRepository.deleteById(accountId);
    }

    @Async
    public CompletableFuture<CustomerModel> findCustomer(String id) {
        return Failsafe.with(retryPolicy(id))
            .getAsync(() -> customerClient.getCustomer(id))
            .exceptionally(err -> null);
    }

    @SuppressWarnings("unchecked")
    private RetryPolicy<Object> retryPolicy(final String id) {
        return new RetryPolicy<>()
            .handle(ConnectException.class, RestClientException.class)
            .withMaxRetries(5)
            .withDelay(Duration.ofMillis(750))
            .onFailedAttempt(obj -> logger.error("Fail to get info from Customer with ID ".concat(id)));
    }

    public void activate(final String id) {
        Optional<Account> account = accountRepository.findById(id);
        if (account.isEmpty()) throw new IllegalArgumentException("Account does not exist with ID.");
        final Account entity = account.get();
        if (entity.isActive()) throw new IllegalArgumentException("Account has been activated.");
        entity.setActive(true);
        accountRepository.save(entity);
    }

    public void updateBalance(final BalanceModel model) {
        Optional<Account> account = accountRepository.findById(model.getAccount());
        if (account.isEmpty()) throw new IllegalArgumentException("Account does not exist with ID.");

        final Account entity = account.get();
        if (!entity.isActive()) throw new IllegalArgumentException("Account is inactive.");

        Double currentBalance = entity.getBalance();
        if ("credit".equals(model.getOperation()))
            currentBalance = currentBalance + model.getValue();
        else if ("debit".equalsIgnoreCase(model.getOperation()))
            currentBalance = currentBalance - model.getValue();

        entity.setBalance(currentBalance);
        accountRepository.save(entity);
    }

}

File: account/src/main/kotlin/br/com/accenture/wallet/account/service/AccountService.kt
package br.com.accenture.wallet.account.service

import br.com.accenture.wallet.account.domain.AccountBean
import br.com.accenture.wallet.account.domain.CustomerBean
import br.com.accenture.wallet.account.repository.AccountRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_SINGLETON
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import java.util.*
import java.util.stream.Collectors

@Service
@Scope(SCOPE_SINGLETON)
class AccountService(val repository: AccountRepository) {
    private val logger = LoggerFactory.getLogger(AccountService::class.java)

    @Value("\${customer.service}")
    private val customerServiceUrl: String? = null

    private val customerClient = RestTemplate()

    fun findAll(): List<AccountBean> {
        return repository.run {
            findAll()
                .stream()
                .map { entity -> AccountBean(null, false).fromEntity(entity) }
                .collect(Collectors.toList())
        }
    }

    fun create(model: AccountBean): AccountBean? {
        val customer = findCustomer(model.customer)
        if (customer.isEmpty()) return null
        logger.info("Criando uma conta para o cliente " + customer.get().name)
        val account = repository.save(model.toEntity())
        model.id = account.id
        return model
    }

    fun remove(accountId: String) {
        repository.deleteById(accountId)
    }

    fun getById(value: String): Optional<AccountBean> {
        val account = repository.findById(value)
        return if (account.isEmpty) Optional.empty() else Optional.of(AccountBean(null, false).fromEntity(account.get()))
    }

    fun findCustomer(id: String): Optional<CustomerBean> {
        return try {
            val customer = customerClient
                .getForEntity(customerServiceUrl + id, CustomerBean::class.java)
                .body
            Optional.of(customer!!)
        } catch (ex: Exception) {
            logger.error(ex.message)
            Optional.empty()
        }

    }
}
